contracts/forwarder/Eip712Forwarder.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier:MIT
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "./IForwarder.sol";
+// TODO now: move getError to the forwarder folder
+import "../utils/GsnUtils.sol";
+
+contract Eip712Forwarder is IForwarder {
+    using ECDSA for bytes32;
+
+    string public constant GENERIC_PARAMS = "address to,bytes data,address from,uint256 nonce,uint256 gas";
+
+    mapping(bytes32 => bool) public typeHashes;
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    function getNonce(address from)
+    public view override
+    returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig)
+    public override view {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+    }
+
+    function execute(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig
+    )
+    public
+    override
+    returns (bool, bytes memory) {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        return req.to.call{gas : req.gas}(abi.encodePacked(req.data, req.from));
```

Comment:
no guarantee that `req.gas` will be passed in here.
Better to handle that in the Forwarder extension. Hence why my comment above about removing the gas parameter from the base forwarder. This is also how I handled in EIP-2585 : https://github.com/wighawag/eip-2585/blob/c9769ad0b4260b4f08f981527bb1162ecce570d3/contracts/src/EIP712Forwarder.sol#L226

Reply:
If we move the gas into the extension, then the forwarder will not be able to apply it to the call, since (by definition) the generic forwarder knows only the generic fields.
The relayer calling the forwarder must have a way to ensure an "upper bound" on the gas limit of the call, and the "execute" is the place to handle that gas limits.
Its true that EVM's gas handling is problematic, and even more when handled in nested calls. That's why it has to be the "leaf" call to enforce the gas limit (that is, it is very hard to estimate the value to apply when calling "forwarder.execute" itself.

Reply:
The current way you apply gas do not protect the signer so I do not see the point of having it there.
Plus you only need to protect the signer if it is the one who pays the gas. hence why it should be outside of the forwarder, that have no need to bother about gas repayment

Gas limit can be easily handled in the forwarder caller as I did in my EIP-1776 demo : https://github.com/wighawag/eip-2585/blob/c9769ad0b4260b4f08f981527bb1162ecce570d3/contracts/src/EIP1776ForwarderWrapper.sol#L102

if better gas handling appears, we can keep using the same forwarder and extension can use better handling